database/redis/notification.go

```diff
@@ -98,7 +130,119 @@ func (connector *DbConnector) removeNotifications(notifications []*moira.Schedul
 }
 
 // FetchNotifications fetch notifications by given timestamp and delete it
-func (connector *DbConnector) FetchNotifications(to int64) ([]*moira.ScheduledNotification, error) {
+func (connector *DbConnector) FetchNotifications(to int64, limit int64) ([]*moira.ScheduledNotification, error) {
+	// No limit
+	if limit == 0 {
+		return connector.fetchNotificationsNoLimit(to)
+	}
+
+	count, err := connector.notificationsCount(to)
+	if err != nil {
+		return nil, err
+	}
+
+	// Hope count will be not greater then limit when we call fetchNotificationsNoLimit
+	if count == nil {
+		return nil, fmt.Errorf("failed to count notification in db")
+	}
+	if *count < limit/2 {
+		return connector.fetchNotificationsNoLimit(to)
+	}
+
+	return connector.fetchNotificationsWithLimit(to, limit)
+}
+
+func (connector *DbConnector) notificationsCount(to int64) (*int64, error) {
+	c := connector.pool.Get()
+	defer c.Close()
+
+	count, err := redis.Int64(c.Do("ZCOUNT", notifierNotificationsKey, "-inf", to))
+	if err != nil {
+		return nil, fmt.Errorf("failed to ZCOUNT to notificationsCount: %s", err)
+	}
+	return &count, nil
+}
+
+// fetchNotificationsWithLimit reads and drops notifications from DB with limit
+func (connector *DbConnector) fetchNotificationsWithLimit(to int64, limit int64) ([]*moira.ScheduledNotification, error) {
+	// fetchNotifecationsWithLimitDo uses WATCH, so transaction may fail and will retry it
+	// see https://redis.io/topics/transactions
+	var e *transactionError
+
+	for i := 0; i < transactionTriesLimit; i++ {
+		res, err := connector.fetchNotificationsWithLimitDo(to, limit)
+
+		if err == nil {
+			return res, nil
+		}
+
+		if errors.As(err, &e) {
```

Comment:
Are you trying to get pointer on pointer? probably this would be better?

```suggestion
		if errors.As(err, &transactionError{}) {
```

Reply:
For some reasons this does not work :(
```./notification.go:187:6: second argument to errors.As must be a pointer to an interface or a type implementing error```